File: portable-core/circuit-lab/CircuitModel.cpp
#include "CircuitModel.h"

#include <cstring>

void CircuitModel::clear() {
    componentCount = 0;
    terminalCount = 0;
    wireCount = 0;
    nextComponentId = 1;
    nextTerminalId = 1;
}

int CircuitModel::addComponent(ComponentType type, int x, int y) {
    if (componentCount >= MaxComponents || terminalCount + 2 > MaxTerminals) {
        return 0;
    }

    Component& component = components[componentCount++];
    component.id = nextComponentId++;
    component.type = type;
    component.x = x;
    component.y = y;

    if (type == ComponentType::Battery) {
        component.terminalIds[0] = addTerminal(component.id, "negative", -32, 0);
        component.terminalIds[1] = addTerminal(component.id, "positive", 32, 0);
    } else {
        component.terminalIds[0] = addTerminal(component.id, "a", -24, 28);
        component.terminalIds[1] = addTerminal(component.id, "b", 24, 28);
    }

    return component.id;
}

bool CircuitModel::addWire(int fromTerminalId, int toTerminalId) {
    if (fromTerminalId == toTerminalId || wireCount >= MaxWires) {
        return false;
    }
    if (findTerminal(fromTerminalId) == nullptr || findTerminal(toTerminalId) == nullptr) {
        return false;
    }
    if (wireExists(fromTerminalId, toTerminalId)) {
        return false;
    }

    wires[wireCount++] = {fromTerminalId, toTerminalId};
    return true;
}

bool CircuitModel::isBulbLit() const {
    for (int i = 0; i < componentCount; ++i) {
        if (components[i].type == ComponentType::Bulb && isBulbLit(components[i].id)) {
            return true;
        }
    }
    return false;
}

bool CircuitModel::isBulbLit(int bulbComponentId) const {
    const Component* bulb = findComponent(bulbComponentId);
    if (bulb == nullptr || bulb->type != ComponentType::Bulb) {
        return false;
    }

    for (int i = 0; i < componentCount; ++i) {
        if (components[i].type == ComponentType::Battery && batteryCanReachThroughBulb(components[i], bulbComponentId)) {
            return true;
        }
    }
    return false;
}

int CircuitModel::getComponentCount() const {
    return componentCount;
}

int CircuitModel::getWireCount() const {
    return wireCount;
}

int CircuitModel::getTerminalId(int componentId, const char* name) const {
    const Component* component = findComponent(componentId);
    if (component == nullptr) {
        return 0;
    }

    const Terminal* terminal = terminalByName(*component, name);
    return terminal ? terminal->id : 0;
}

int CircuitModel::addTerminal(int componentId, const char* name, int localX, int localY) {
    if (terminalCount >= MaxTerminals) {
        return 0;
    }

    Terminal& terminal = terminals[terminalCount++];
    terminal.id = nextTerminalId++;
    terminal.componentId = componentId;
    terminal.name = name;
    terminal.localX = localX;
    terminal.localY = localY;
    return terminal.id;
}

const Component* CircuitModel::findComponent(int componentId) const {
    for (int i = 0; i < componentCount; ++i) {
        if (components[i].id == componentId) {
            return &components[i];
        }
    }
    return nullptr;
}

const Terminal* CircuitModel::findTerminal(int terminalId) const {
    for (int i = 0; i < terminalCount; ++i) {
        if (terminals[i].id == terminalId) {
            return &terminals[i];
        }
    }
    return nullptr;
}

const Terminal* CircuitModel::terminalByName(const Component& component, const char* name) const {
    for (int i = 0; i < 2; ++i) {
        const Terminal* terminal = findTerminal(component.terminalIds[i]);
        if (terminal != nullptr && terminalNameEquals(*terminal, name)) {
            return terminal;
        }
    }
    return nullptr;
}

bool CircuitModel::wireExists(int a, int b) const {
    for (int i = 0; i < wireCount; ++i) {
        if ((wires[i].fromTerminalId == a && wires[i].toTerminalId == b) ||
            (wires[i].fromTerminalId == b && wires[i].toTerminalId == a)) {
            return true;
        }
    }
    return false;
}

bool CircuitModel::batteryCanReachThroughBulb(const Component& battery, int bulbComponentId) const {
    const Terminal* positive = terminalByName(battery, "positive");
    const Terminal* negative = terminalByName(battery, "negative");
    if (positive == nullptr || negative == nullptr) {
        return false;
    }

    GraphEdge graph[MaxTerminals][MaxTerminals] = {};
    int edgeCounts[MaxTerminals] = {};

    for (int i = 0; i < wireCount; ++i) {
        addGraphEdge(graph, edgeCounts, wires[i].fromTerminalId, wires[i].toTerminalId, 0);
    }

    for (int i = 0; i < componentCount; ++i) {
        const Component& component = components[i];
        if (component.type == ComponentType::Bulb) {
            addGraphEdge(graph, edgeCounts, component.terminalIds[0], component.terminalIds[1], component.id);
        }
    }

    struct SearchState {
        int terminalId;
        bool usedTargetBulb;
    };

    SearchState queue[MaxTerminals * 2] = {};
    bool visited[MaxTerminals][2] = {};
    int head = 0;
    int tail = 0;

    const int positiveIndex = terminalIndex(positive->id);
    if (positiveIndex < 0) {
        return false;
    }

    queue[tail++] = {positive->id, false};
    visited[positiveIndex][0] = true;

    while (head < tail) {
        const SearchState state = queue[head++];
        if (state.terminalId == negative->id && state.usedTargetBulb) {
            return true;
        }

        const int fromIndex = terminalIndex(state.terminalId);
        if (fromIndex < 0) {
            continue;
        }

        for (int i = 0; i < edgeCounts[fromIndex]; ++i) {
            const GraphEdge& edge = graph[fromIndex][i];
            const bool nextUsed = state.usedTargetBulb || edge.bulbComponentId == bulbComponentId;
            const int nextIndex = terminalIndex(edge.toTerminalId);
            if (nextIndex >= 0 && !visited[nextIndex][nextUsed ? 1 : 0] && tail < MaxTerminals * 2) {
                visited[nextIndex][nextUsed ? 1 : 0] = true;
                queue[tail++] = {edge.toTerminalId, nextUsed};
            }
        }
    }

    return false;
}

bool CircuitModel::terminalNameEquals(const Terminal& terminal, const char* name) const {
    return std::strcmp(terminal.name, name) == 0;
}

void CircuitModel::addGraphEdge(
    GraphEdge graph[MaxTerminals][MaxTerminals],
    int edgeCounts[MaxTerminals],
    int from,
    int to,
    int bulbComponentId
) const {
    const int fromIndex = terminalIndex(from);
    const int toIndex = terminalIndex(to);
    if (fromIndex < 0 || toIndex < 0) {
        return;
    }
    if (edgeCounts[fromIndex] < MaxTerminals) {
        graph[fromIndex][edgeCounts[fromIndex]++] = {to, bulbComponentId};
    }
    if (edgeCounts[toIndex] < MaxTerminals) {
        graph[toIndex][edgeCounts[toIndex]++] = {from, bulbComponentId};
    }
}

int CircuitModel::terminalIndex(int terminalId) const {
    for (int i = 0; i < terminalCount; ++i) {
        if (terminals[i].id == terminalId) {
            return i;
        }
    }
    return -1;
}

File: desktop-prototypes/circuit-lab-python/circuit_model.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field
from enum import Enum
from itertools import count
from math import hypot


SNAP_DISTANCE = 15


class ComponentType(Enum):
    BATTERY = "battery"
    BULB = "bulb"


@dataclass
class Terminal:
    id: int
    component_id: int
    name: str
    local_x: int
    local_y: int


@dataclass
class Component:
    id: int
    type: ComponentType
    x: int
    y: int
    terminals: list[Terminal] = field(default_factory=list)


@dataclass
class Wire:
    from_terminal_id: int
    to_terminal_id: int


class CircuitModel:
    def __init__(self) -> None:
        self.components: list[Component] = []
        self.wires: list[Wire] = []
        self._component_ids = count(1)
        self._terminal_ids = count(1)

    def clear(self) -> None:
        self.components.clear()
        self.wires.clear()
        self._component_ids = count(1)
        self._terminal_ids = count(1)

    def add_component(self, component_type: ComponentType | str, x: int, y: int) -> Component:
        normalized_type = self._normalize_component_type(component_type)
        component_id = next(self._component_ids)
        component = Component(component_id, normalized_type, x, y)

        if normalized_type == ComponentType.BATTERY:
            component.terminals = [
                self._new_terminal(component_id, "negative", -32, 0),
                self._new_terminal(component_id, "positive", 32, 0),
            ]
        elif normalized_type == ComponentType.BULB:
            component.terminals = [
                self._new_terminal(component_id, "a", -24, 28),
                self._new_terminal(component_id, "b", 24, 28),
            ]

        self.components.append(component)
        return component

    def add_wire(self, from_terminal_id: int, to_terminal_id: int) -> bool:
        if from_terminal_id == to_terminal_id:
            return False
        if self._find_terminal(from_terminal_id) is None or self._find_terminal(to_terminal_id) is None:
            return False
        if self._wire_exists(from_terminal_id, to_terminal_id):
            return False

        self.wires.append(Wire(from_terminal_id, to_terminal_id))
        return True

    def move_component(self, component_id: int, x: int, y: int) -> None:
        component = self.get_component(component_id)
        if component:
            component.x = x
            component.y = y

    def get_component(self, component_id: int) -> Component | None:
        return next((component for component in self.components if component.id == component_id), None)

    def get_terminal_world_pos(self, terminal_id: int) -> tuple[int, int] | None:
        terminal = self._find_terminal(terminal_id)
        if terminal is None:
            return None

        component = self.get_component(terminal.component_id)
        if component is None:
            return None

        return component.x + terminal.local_x, component.y + terminal.local_y

    def find_nearest_terminal(self, x: int, y: int, max_distance: int = SNAP_DISTANCE) -> Terminal | None:
        best_terminal = None
        best_distance = max_distance

        for component in self.components:
            for terminal in component.terminals:
                terminal_pos = self.get_terminal_world_pos(terminal.id)
                if terminal_pos is None:
                    continue
                terminal_x, terminal_y = terminal_pos
                distance = hypot(terminal_x - x, terminal_y - y)
                if distance <= best_distance:
                    best_distance = distance
                    best_terminal = terminal

        return best_terminal

    def component_at(self, x: int, y: int) -> Component | None:
        for component in reversed(self.components):
            if component.type == ComponentType.BATTERY:
                if component.x - 40 <= x <= component.x + 40 and component.y - 22 <= y <= component.y + 22:
                    return component
            elif component.type == ComponentType.BULB:
                if hypot(component.x - x, component.y - y) <= 34:
                    return component
        return None

    def lit_bulb_ids(self) -> set[int]:
        lit: set[int] = set()
        batteries = [component for component in self.components if component.type == ComponentType.BATTERY]
        bulbs = [component for component in self.components if component.type == ComponentType.BULB]

        for bulb in bulbs:
            for battery in batteries:
                if self._battery_can_reach_through_bulb(battery, bulb.id):
                    lit.add(bulb.id)
                    break

        return lit

    def is_bulb_lit(self, bulb_id: int | None = None) -> bool:
        lit = self.lit_bulb_ids()
        if bulb_id is None:
            return bool(lit)
        return bulb_id in lit

    def get_status_text(self) -> str:
        if self.is_bulb_lit():
            return "Circuit complete: bulb ON"
        if self.components or self.wires:
            return "Circuit incomplete"
        return "Select Battery/Bulb/Wire"

    def _battery_can_reach_through_bulb(self, battery: Component, bulb_id: int) -> bool:
        positive = self._terminal_by_name(battery, "positive")
        negative = self._terminal_by_name(battery, "negative")
        if positive is None or negative is None:
            return False

        graph = self._build_graph()
        queue = deque([(positive.id, False)])
        visited = {(positive.id, False)}

        while queue:
            terminal_id, used_target_bulb = queue.popleft()
            if terminal_id == negative.id and used_target_bulb:
                return True

            for next_terminal_id, edge_bulb_id in graph.get(terminal_id, []):
                next_used = used_target_bulb or edge_bulb_id == bulb_id
                state = (next_terminal_id, next_used)
                if state not in visited:
                    visited.add(state)
                    queue.append(state)

        return False

    def _build_graph(self) -> dict[int, list[tuple[int, int | None]]]:
        graph: dict[int, list[tuple[int, int | None]]] = {}

        def add_edge(a: int, b: int, bulb_id: int | None = None) -> None:
            graph.setdefault(a, []).append((b, bulb_id))
            graph.setdefault(b, []).append((a, bulb_id))

        for wire in self.wires:
            add_edge(wire.from_terminal_id, wire.to_terminal_id)

        for component in self.components:
            if component.type == ComponentType.BULB and len(component.terminals) == 2:
                add_edge(component.terminals[0].id, component.terminals[1].id, component.id)

        return graph

    def _new_terminal(self, component_id: int, name: str, local_x: int, local_y: int) -> Terminal:
        return Terminal(next(self._terminal_ids), component_id, name, local_x, local_y)

    def _find_terminal(self, terminal_id: int) -> Terminal | None:
        for component in self.components:
            for terminal in component.terminals:
                if terminal.id == terminal_id:
                    return terminal
        return None

    def _terminal_by_name(self, component: Component, name: str) -> Terminal | None:
        return next((terminal for terminal in component.terminals if terminal.name == name), None)

    def _wire_exists(self, a: int, b: int) -> bool:
        return any(
            {wire.from_terminal_id, wire.to_terminal_id} == {a, b}
            for wire in self.wires
        )

    def _normalize_component_type(self, component_type: ComponentType | str) -> ComponentType:
        if isinstance(component_type, ComponentType):
            return component_type
        try:
            return ComponentType(component_type)
        except ValueError as error:
            raise ValueError(f"Unsupported component type: {component_type}") from error
